### PR TITLE
mysql-compatibility: Add GROUP BY implicit order

### PR DESCRIPTION
**Pull request by Joyinqin**
### What is changed, added or deleted? (Required)
TiDB differs from MySQL in a way that might impact migrations. Here is a testcase to prove it:

```
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 (
    id INT NOT NULL PRIMARY KEY auto_increment,
    b INT NOT NULL
);
INSERT INTO t1 (b) VALUES (1),(2),(2),(3),(1),(4);
SELECT b, COUNT(*) FROM t1 GROUP BY b;

TiDB:

tidb> SELECT b, COUNT(*) FROM t1 GROUP BY b;
+---+----------+
| b | COUNT(*) |
+---+----------+
| 3 |        1 |
| 2 |        2 |
| 4 |        1 |
| 1 |        2 |
+---+----------+
4 rows in set (0.00 sec)

MySQL:

mysql [localhost:5731] {msandbox} (test) > SELECT b, COUNT(*) FROM t1 GROUP BY b;
+---+----------+
| b | COUNT(*) |
+---+----------+
| 1 |        2 |
| 2 |        2 |
| 3 |        1 |
| 4 |        1 |
+---+----------+
4 rows in set (0.00 sec)

Note the difference in plan showing this is intentionally sorted:

mysql [localhost:5731] {msandbox} (test) > explain SELECT b, COUNT(*) FROM t1 GROUP BY b;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra                           |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
|  1 | SIMPLE      | t1    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    6 |   100.00 | Using temporary; Using filesort |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+---------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql [localhost:5731] {msandbox} (test) > explain SELECT b, COUNT(*) FROM t1 GROUP BY b ORDER BY NULL;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-----------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra           |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-----------------+
|  1 | SIMPLE      | t1    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    6 |   100.00 | Using temporary |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-----------------+
1 row in set, 1 warning (0.00 sec)
```

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [x] v3.1 (TiDB 3.1 versions)
- [x] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/4437
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by Joyinqin** [comment]
/label needs-cherry-pick-3.0,
needs-cherry-pick-3.1,
needs-cherry-pick-4.0, translation/from-docs, size/small

File: mysql-compatibility.md [hunk 1]
```diff
@@ -117,6 +117,7 @@ TiDB 中的[信息统计](/statistics.md#手动收集) 与 MySQL 中的有所不
 
 - 不支持 `SELECT ... INTO @变量` 语法。
 - 不支持 `SELECT ... GROUP BY ... WITH ROLLUP` 语法。
+- TiDB 中的 `SELECT .. GROUP BY expr` 语法与 MySQL 5.7 中的 `GROUP BY expr ORDER BY expr` 并不一致。TiDB 中的该语法并不代表默认顺序，反而更符合 MySQL 8.0 的行为。
```
Comment (TomShawn) [inline mysql-compatibility.md:120]:
```suggestion
- TiDB 中的 `SELECT .. GROUP BY expr` 语法与 MySQL 5.7 中的 `GROUP BY expr ORDER BY expr` 并不一致。TiDB 中的该语法并不代表默认顺序，而是与 MySQL 8.0 的行为一致。
```

File: mysql-compatibility.md [hunk 2]
```diff
@@ -117,6 +117,7 @@ TiDB 中的[信息统计](/statistics.md#手动收集) 与 MySQL 中的有所不
 
 - 不支持 `SELECT ... INTO @变量` 语法。
 - 不支持 `SELECT ... GROUP BY ... WITH ROLLUP` 语法。
+- TiDB 中的 `SELECT .. GROUP BY expr` 语法与 MySQL 5.7 中的 `GROUP BY expr ORDER BY expr` 并不一致。TiDB 中的该语法并不代表默认顺序，而是与 MySQL 8.0 的行为一致。
```
Comment (TomShawn) [inline mysql-compatibility.md:120]:
```suggestion
- TiDB 中的 `SELECT .. GROUP BY expr` 语法与 MySQL 5.7 中的 `GROUP BY expr ORDER BY expr` 并不一致。TiDB 中该语法并不代表默认顺序，而是与 MySQL 8.0 的行为一致。
```

**Comment by TomShawn** [comment]
LGTM

File: mysql-compatibility.md [hunk 3]
```diff
@@ -116,6 +116,7 @@ TiDB 中的[信息统计](/statistics.md#手动收集) 与 MySQL 中的有所不
 
 - 不支持 `SELECT ... INTO @变量` 语法。
 - 不支持 `SELECT ... GROUP BY ... WITH ROLLUP` 语法。
+- TiDB 中的 `SELECT .. GROUP BY expr` 语法与 MySQL 5.7 中的 `GROUP BY expr ORDER BY expr` 并不一致。TiDB 中该语法并不代表默认顺序，而是与 MySQL 8.0 的行为一致。
```
Comment (zz-jason) [inline mysql-compatibility.md:119]:
```suggestion
- TiDB 中的 `SELECT .. GROUP BY expr` 的返回结果与 MySQL 5.7 并不一致。MySQL 5.7 的结果等价于 `GROUP BY expr ORDER BY expr`，TiDB 中该语法所返回的结果并不承诺任何顺序，与 MySQL 8.0 的行为一致。
```

**Review by TomShawn — APPROVED** [review]

**Pull request merged**